File: assignment 9/python/programs/8.py
def group_elements_by_first_occurrence(arr):
    element_indices = {}  # Dictionary to store the first occurrence indices of elements
    grouped_elements = {}

    for index, element in enumerate(arr):
        if element in element_indices:
            first_occurrence_index = element_indices[element]
            if element in grouped_elements:
                grouped_elements[element].append(
                    arr[first_occurrence_index : index + 1]
                )
            else:
                grouped_elements[element] = [arr[first_occurrence_index : index + 1]]
        else:
            element_indices[element] = index

    return list(grouped_elements.values())


# Example usage
arr = [1, 2, 3, 1, 2, 4, 5, 3, 6]
grouped = group_elements_by_first_occurrence(arr)
print("Grouped elements based on their first occurrence:")
for group in grouped:
    print(group)

File: assignment 9/c/programs/8.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void groupElementsByFirstOccurrence(int arr[], int size)
{
    // Create a dictionary to store the first occurrence index of each element
    int *firstOccurrence = (int *)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++)
    {
        firstOccurrence[i] = -1; // Initialize with -1 to indicate no occurrence yet
    }

    // Create a counter for the group numbers
    int groupCounter = 0;

    for (int i = 0; i < size; i++)
    {
        if (firstOccurrence[arr[i]] == -1)
        {
            // This element is seen for the first time
            firstOccurrence[arr[i]] = groupCounter++;
            printf("Group %d: %d", firstOccurrence[arr[i]], arr[i]);
        }
        else
        {
            // This element has been seen before
            printf(" (Group %d)", firstOccurrence[arr[i]]);
        }

        printf("\n");
    }

    free(firstOccurrence);
}

int main()
{
    int arr[] = {4, 2, 4, 1, 6, 2, 4};
    int size = sizeof(arr) / sizeof(arr[0]);

    printf("Grouped elements based on their first occurrence:\n");
    groupElementsByFirstOccurrence(arr, size);

    return 0;
}
